File: api/routers/doctor.py
"""Doctor use cases: View Patient Records (auto-includes AI prediction results),
Review Prediction Results, Generate Clinical Recommendation, Download Prediction
Report."""

from __future__ import annotations

from typing import Optional

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import Response
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from api.auth import log_activity, require_role
from api.db import get_db
from api.models import ClinicalRecommendation, HealthRecord, UploadedReport, User
from api.services import build_pdf_report, record_to_dict

router = APIRouter(prefix="/api/doctor", tags=["doctor"])

doctor_only = require_role("doctor")


class ReviewRequest(BaseModel):
    note: str = Field(..., min_length=3)


class RecommendationRequest(BaseModel):
    recommendation: str = Field(..., min_length=3)
    record_id: Optional[int] = None


@router.get("/patients")
def list_patients(user: User = Depends(doctor_only), db: Session = Depends(get_db)) -> list[dict]:
    patients = db.query(User).filter(User.role == "patient").order_by(User.full_name).all()
    counts = {
        p.id: db.query(HealthRecord).filter(HealthRecord.patient_id == p.id).count()
        for p in patients
    }
    latest = {}
    for p in patients:
        rec = (
            db.query(HealthRecord)
            .filter(HealthRecord.patient_id == p.id)
            .order_by(HealthRecord.created_at.desc())
            .first()
        )
        latest[p.id] = rec
    return [
        {
            "id": p.id,
            "full_name": p.full_name,
            "email": p.email,
            "assessments": counts[p.id],
            "latest_risk": latest[p.id].risk_classification if latest[p.id] else None,
            "latest_probability": latest[p.id].risk_probability if latest[p.id] else None,
        }
        for p in patients
    ]


@router.get("/patients/{patient_id}/records")
def view_patient_records(
    patient_id: int,
    user: User = Depends(doctor_only),
    db: Session = Depends(get_db),
) -> dict:
    """'View Patient Records' — per the <<include>> relationship, the patient's
    AI prediction results are aggregated directly into the record view."""
    patient = db.get(User, patient_id)
    if patient is None or patient.role != "patient":
        raise HTTPException(404, "Patient not found.")

    records = (
        db.query(HealthRecord)
        .filter(HealthRecord.patient_id == patient_id)
        .order_by(HealthRecord.created_at.desc())
        .all()
    )
    uploads = (
        db.query(UploadedReport)
        .filter(UploadedReport.patient_id == patient_id)
        .order_by(UploadedReport.uploaded_at.desc())
        .all()
    )
    recommendations = (
        db.query(ClinicalRecommendation)
        .filter(ClinicalRecommendation.patient_id == patient_id)
        .order_by(ClinicalRecommendation.created_at.desc())
        .all()
    )
    log_activity(db, user, "view_patient_records", f"patient={patient_id}")
    return {
        "patient": {
            "id": patient.id,
            "full_name": patient.full_name,
            "email": patient.email,
            "phone": patient.phone,
            "date_of_birth": patient.date_of_birth,
        },
        "records": [record_to_dict(r) for r in records],
        "uploads": [
            {"id": u.id, "filename": u.filename, "uploaded_at": u.uploaded_at.isoformat()}
            for u in uploads
        ],
        "recommendations": [
            {
                "id": rec.id,
                "recommendation": rec.recommendation,
                "created_at": rec.created_at.isoformat(),
            }
            for rec in recommendations
        ],
    }


@router.post("/records/{record_id}/review")
def review_prediction(
    record_id: int,
    payload: ReviewRequest,
    user: User = Depends(doctor_only),
    db: Session = Depends(get_db),
) -> dict:
    """'Review Prediction Results'."""
    record = db.get(HealthRecord, record_id)
    if record is None:
        raise HTTPException(404, "Record not found.")
    record.reviewed_by_id = user.id
    record.review_note = payload.note
    record.reviewed_at = datetime.utcnow()
    db.add(record)
    db.commit()
    log_activity(db, user, "review_prediction", f"record={record_id}")
    return {"message": "Review saved."}


@router.post("/patients/{patient_id}/recommendations")
def generate_recommendation(
    patient_id: int,
    payload: RecommendationRequest,
    user: User = Depends(doctor_only),
    db: Session = Depends(get_db),
) -> dict:
    """'Generate Clinical Recommendation'."""
    patient = db.get(User, patient_id)
    if patient is None or patient.role != "patient":
        raise HTTPException(404, "Patient not found.")
    rec = ClinicalRecommendation(
        patient_id=patient_id,
        doctor_id=user.id,
        record_id=payload.record_id,
        recommendation=payload.recommendation,
    )
    db.add(rec)
    db.commit()
    log_activity(db, user, "generate_recommendation", f"patient={patient_id}")
    return {"message": "Recommendation saved.", "id": rec.id}


@router.get("/records/{record_id}/report")
def download_report(
    record_id: int,
    user: User = Depends(doctor_only),
    db: Session = Depends(get_db),
) -> Response:
    """'Download Prediction Report' (doctor side, e.g. for EHR filing)."""
    record = db.get(HealthRecord, record_id)
    if record is None:
        raise HTTPException(404, "Record not found.")
    patient = db.get(User, record.patient_id)
    pdf_bytes = build_pdf_report(db, record, patient)
    log_activity(db, user, "download_report", f"record={record_id}")
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={
            "Content-Disposition": f'attachment; filename="cvd_report_{record_id}.pdf"'
        },
    )

File: server/src/routes/doctor.js
import { Router } from "express";
import { requireRole } from "../auth.js";
import { ClinicalRecommendation, DoctorNote, HealthRecord, Upload, User, logActivity } from "../models.js";
import { buildPdfReport } from "../pdf.js";
import { recordPayload } from "../serialize.js";

const router = Router();
const doctorOnly = requireRole("doctor");

// ---- Doctor notes on a patient ------------------------------------------------

router.get("/patients/:id/notes", doctorOnly, async (req, res) => {
  const notes = await DoctorNote.find({ patient: req.params.id })
    .sort({ createdAt: -1 })
    .populate("doctor", "fullName");
  res.json(notes.map((n) => ({
    id: String(n._id),
    note: n.note,
    doctor: n.doctor?.fullName || "Doctor",
    mine: String(n.doctor?._id) === String(req.user._id),
    created_at: n.createdAt.toISOString(),
  })));
});

router.post("/patients/:id/notes", doctorOnly, async (req, res) => {
  const patient = await User.findOne({ _id: req.params.id, role: "patient" });
  if (!patient) return res.status(404).json({ detail: "Patient not found." });
  const note = String(req.body?.note || "").trim();
  if (note.length < 3) return res.status(422).json({ detail: "Note is too short." });
  const doc = await DoctorNote.create({ patient: patient._id, doctor: req.user._id, note });
  await logActivity(req.user, "add_doctor_note", `patient=${patient._id}`);
  res.json({ message: "Note added.", id: String(doc._id) });
});

router.delete("/notes/:id", doctorOnly, async (req, res) => {
  const note = await DoctorNote.findOne({ _id: req.params.id, doctor: req.user._id });
  if (!note) return res.status(404).json({ detail: "Note not found (you can only delete your own notes)." });
  await note.deleteOne();
  res.json({ message: "Note deleted." });
});

router.get("/patients", doctorOnly, async (req, res) => {
  const patients = await User.find({ role: "patient" }).sort({ fullName: 1 });
  const out = [];
  for (const p of patients) {
    const latest = await HealthRecord.findOne({ patient: p._id }).sort({ createdAt: -1 });
    const count = await HealthRecord.countDocuments({ patient: p._id });
    out.push({
      id: String(p._id),
      patient_id: p.patientId,
      full_name: p.fullName,
      email: p.email,
      assessments: count,
      latest_risk: latest ? latest.riskClassification : null,
      latest_risk_level: latest ? latest.riskLevel : null,
      latest_probability: latest ? latest.riskProbability : null,
    });
  }
  res.json(out);
});

router.get("/patients/:id/records", doctorOnly, async (req, res) => {
  const patient = await User.findById(req.params.id).catch(() => null);
  if (!patient || patient.role !== "patient")
    return res.status(404).json({ detail: "Patient not found." });

  const records = await HealthRecord.find({ patient: patient._id }).sort({ createdAt: -1 });
  const uploads = await Upload.find({ patient: patient._id }).sort({ uploadedAt: -1 });
  const recs = await ClinicalRecommendation.find({ patient: patient._id }).sort({ createdAt: -1 });
  await logActivity(req.user, "view_patient_records", `patient=${patient._id}`);

  res.json({
    patient: {
      id: String(patient._id),
      patient_id: patient.patientId,
      full_name: patient.fullName,
      email: patient.email,
      phone: patient.phone || null,
      date_of_birth: patient.dateOfBirth || null,
      medical_history: patient.medicalHistory,
    },
    records: records.map(recordPayload),
    uploads: uploads.map((u) => ({
      id: String(u._id), filename: u.filename, uploaded_at: u.uploadedAt.toISOString(),
    })),
    recommendations: recs.map((r) => ({
      id: String(r._id), recommendation: r.recommendation, created_at: r.createdAt.toISOString(),
    })),
  });
});

router.post("/records/:id/review", doctorOnly, async (req, res) => {
  const note = req.body?.note;
  if (!note || note.length < 3) return res.status(422).json({ detail: "Review note is too short." });
  const record = await HealthRecord.findById(req.params.id).catch(() => null);
  if (!record) return res.status(404).json({ detail: "Record not found." });
  record.reviewedBy = req.user._id;
  record.reviewNote = note;
  record.reviewedAt = new Date();
  await record.save();
  await logActivity(req.user, "review_prediction", `record=${record._id}`);
  res.json({ message: "Review saved." });
});

router.post("/patients/:id/recommendations", doctorOnly, async (req, res) => {
  const text = req.body?.recommendation;
  if (!text || text.length < 3) return res.status(422).json({ detail: "Recommendation is too short." });
  const patient = await User.findById(req.params.id).catch(() => null);
  if (!patient || patient.role !== "patient")
    return res.status(404).json({ detail: "Patient not found." });
  const rec = await ClinicalRecommendation.create({
    patient: patient._id,
    doctor: req.user._id,
    record: req.body?.record_id || undefined,
    recommendation: text,
  });
  await logActivity(req.user, "generate_recommendation", `patient=${patient._id}`);
  res.json({ message: "Recommendation saved.", id: String(rec._id) });
});

router.get("/records/:id/report", doctorOnly, async (req, res) => {
  const record = await HealthRecord.findById(req.params.id).catch(() => null);
  if (!record) return res.status(404).json({ detail: "Record not found." });
  const patient = await User.findById(record.patient);
  const recs = await ClinicalRecommendation.find({ patient: record.patient })
    .sort({ createdAt: -1 }).limit(5);
  const pdf = await buildPdfReport(record, patient, recs);
  await logActivity(req.user, "download_report", `record=${record._id}`);
  res.set({
    "Content-Type": "application/pdf",
    "Content-Disposition": `attachment; filename="cvd_report_${record._id}.pdf"`,
  }).send(pdf);
});

export default router;
